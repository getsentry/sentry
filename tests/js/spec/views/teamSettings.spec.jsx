import React from 'react';
import {mountWithTheme} from 'sentry-test/enzyme';

import TeamStore from 'app/stores/teamStore';
import TeamSettings from 'app/views/settings/organizationTeams/teamSettings';

describe('TeamSettings', function() {
  beforeEach(function() {
    MockApiClient.clearMockResponses();
    jest.spyOn(window.location, 'assign');
  });

  afterEach(function() {
    window.location.assign.mockRestore();
  });

  it('can change slug', async function() {
    const team = TestStubs.Team();
    const putMock = MockApiClient.addMockResponse({
      url: `/teams/org/${team.slug}/`,
      method: 'PUT',
    });
    const mountOptions = TestStubs.routerContext();
    const {router} = mountOptions.context;

    const wrapper = mountWithTheme(
      <TeamSettings
        routes={[]}
        router={router}
        params={{orgId: 'org', teamId: team.slug}}
        team={team}
        onTeamChange={() => {}}
      />,
      mountOptions
    );

    wrapper
      .find('input[name="slug"]')
      .simulate('change', {target: {value: 'NEW SLUG'}})
      .simulate('blur');

    wrapper.find('SaveButton').simulate('click');

    expect(putMock).toHaveBeenCalledWith(
      `/teams/org/${team.slug}/`,
      expect.objectContaining({
        data: {
          slug: 'new-slug',
        },
      })
    );

    await tick();
    expect(router.replace).toHaveBeenCalledWith('/settings/org/teams/new-slug/settings/');
  });

  it('needs team:admin in order to see an enabled Remove Team button', function() {
    const team = TestStubs.Team();

    const wrapper = mountWithTheme(
      <TeamSettings
        routes={[]}
        params={{orgId: 'org', teamId: team.slug}}
        team={team}
        onTeamChange={() => {}}
      />,
      TestStubs.routerContext([{organization: TestStubs.Organization({access: []})}])
    );
    expect(
      wrapper
        .find('Panel')
        .last()
        .find('Button')
        .prop('disabled')
    ).toBe(true);
  });

  it('can remove team', async function() {
    const team = TestStubs.Team({hasAccess: true});
    const deleteMock = MockApiClient.addMockResponse({
      url: `/teams/org/${team.slug}/`,
      method: 'DELETE',
    });
    const routerPushMock = jest.fn();
    jest.spyOn(TeamStore, 'trigger');
    TeamStore.loadInitialData([
      {
        slug: 'team-slug',
        hasAccess: true,
      },
    ]);

    const wrapper = mountWithTheme(
      <TeamSettings
        router={{replace: routerPushMock}}
        routes={[]}
        params={{orgId: 'org', teamId: team.slug}}
        team={team}
        onTeamChange={() => {}}
      />,
      TestStubs.routerContext()
    );

    // Click "Remove Team button
    wrapper.find('Button[priority="danger"] button').simulate('click');

    TeamStore.trigger.mockReset();

    // Wait for modal
    wrapper.find('ModalDialog Button[priority="danger"] button').simulate('click');
    expect(deleteMock).toHaveBeenCalledWith(
      `/teams/org/${team.slug}/`,
      expect.objectContaining({
        method: 'DELETE',
      })
    );

    await tick();
    await tick();
    expect(routerPushMock).toHaveBeenCalledWith('/settings/org/teams/');

    expect(TeamStore.items).toEqual([]);

    TeamStore.trigger.mockRestore();
  });
});
