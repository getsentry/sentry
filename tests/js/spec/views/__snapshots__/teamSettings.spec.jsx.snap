// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TeamSettings render() renders 1`] = `
<SideEffect(DocumentTitle)
  title="Team Settings"
>
  <div
    className="box"
  >
    <div
      className="box-content with-padding"
    >
      <ApiForm
        apiEndpoint="/teams/org/team-slug/"
        apiMethod="PUT"
        cancelLabel="Cancel"
        className="form-stacked"
        errorMessage="Unable to save your changes. Please ensure all fields are valid and try again."
        footerClass="form-actions align-right"
        hideErrors={false}
        initialData={
          Object {
            "name": "Team Name",
            "slug": "team-slug",
          }
        }
        onSubmitSuccess={[Function]}
        requireChanges={true}
        resetOnError={false}
        submitDisabled={false}
        submitLabel="Save Changes"
      >
        <TextField
          disabled={false}
          hideErrorMessage={false}
          label="Name"
          name="name"
          placeholder="e.g. API Team"
          required={true}
        />
        <TextField
          disabled={false}
          hideErrorMessage={false}
          label="Short name"
          name="slug"
          placeholder="e.g. api-team"
          required={true}
        />
      </ApiForm>
    </div>
  </div>
</SideEffect(DocumentTitle)>
`;
